rfd/0143-external-k8s-joining.md

```diff
@@ -0,0 +1,280 @@
+      cluster: my-other-cluster
+```
+
+This configuration is defined in Protobuf form as:
+
+```protobuf
+syntax = "proto3";
+message ProvisionTokenSpecV2KubernetesJWKS {
+  // Rule is a set of properties the Kubernetes-issued token might have to be
+  // allowed to use this ProvisionToken
+  message Rule {
+    // ServiceAccount is the namespaced name of the Kubernetes service account.
+    // Its format is "namespace:service-account".
+    string ServiceAccount = 1;
+  }
+
+  // Cluster is a set of properties that describe a cluster that an entity may
+  // attempt to join from.
+  message Cluster {
+    // Name is an identifier configured for the cluster. This value is chosen
+    // by the user, and does not have to match a specific value configured in
+    // Kubernetes.
+    string Name = 1;
+    // Source is how the public signing keys for the cluster will be obtained
+    // to use for validating a provided JWT.
+    oneof Source {
+      // StaticJWKS allows the public signing keys for a cluster to be
+      // specified statically with the JSON body of the JWKS endpoint.
+      string StaticJWKS = 2;
+    }
+  }
+  
+  // Clusters is a list of Cluster that are accepted sources of JWTs to use
+  // against this token. A Service Account JWT signed by a cluster not listed
+  // here will be rejected.
+  repeated Cluster Clusters = 1;
+  // Allow is a list of Rules, nodes using this token must match one
+  // allow rule to use this token.
+  repeated Rule Allow = 2;
+}
+```
+
+In order to obtain the `jwks` value for their Kubernetes Cluster, operators
+will need to run the following steps from a machine already authenticated
+with their cluster:
+
+- `kubectl proxy -p 8080`
+- `curl http://localhost:8080/openid/v1/jwks`
+
+In the case that an operator rotates the certificate authority used by their
+Kubernetes Cluster to sign Service Account JWTS, the `jwks` field will also
+need to be updated.
+
+#### Configuring a pod for `kubernetes-jwks` joining
+
+The Pod will require a service account with a role granting it the ability to
+call the TokenRequest endpoint for itself. This allows the creation of a token
+with an audience of our choosing in order to include a nonce during the join
+to prevent token reuse.
+
+E.g for a service account named: "argocd":
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: ClusterRole
+metadata:
+  name: argocd-token-request
+rules:
+- apiGroups: [""]
+  resources:
+  - "serviceaccounts/token"
+  verbs:
+  - "create"
+  # Only grant the ability to create tokens for itself. Missing this field leads
+  # to a dangerously powerful role.
+  resourceNames:
+    - argocd
+---
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: RoleBinding
+metadata:
+  name: argocd-token-review
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: argocd-token-review
+subjects:
+- kind: ServiceAccount
+  name: argocd
+  namespace: default
+```
+
```

Comment:
For teleport plugins and `teleport-kube-agent` we can offer a Helm chart, the user would not have to create those resources.

Reply:
I'll add words to this effect - I thought it was important to include this in the RFD as we do need this RBAC grant to exist or we won't be able to call `TokenRequest`.

Reply:
6e07820a3a